sdk/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java

```diff
@@ -0,0 +1,70 @@
+package io.opentelemetry.sdk.common.export;
+
+import java.util.concurrent.Executor;
+
+/**
+ * The implementation of Export operations are often asynchronous in nature, hence the need to
+ * convey a result at a later time. CompletableResultCode facilitates this.
+ *
+ * <p>This class models JDK 8's CompletableFuture to afford migration should Open Telemetry's SDK
+ * select JDK 8 or greater as a baseline, and also to offer familiarity to developers.
+ */
+public class CompletableResultCode {
+  public CompletableResultCode() {}
+
+  private volatile boolean succeeded = false;
+  private boolean completed = false;
+  private Runnable action = null;
+  private Executor actionExecutor = null;
+
+  /** The export operation finished successfully. */
+  public synchronized void success() {
```

Comment:
I'm not super happy with having a public method be synchronized. There is an unintended consequence to this that if someone externally uses an instance of this class to synchronize on, it will lock this from every executing.

The combination of volatile and public synchronized methods makes me a bit queasy. Can we use some more modern concurrency primitives, like Atomics to make this a little less prone to accidental (or intentional) abuse?

At the very least, if we do need to have old-school synchronization in here, we should synchronize on a private final field that is not accessible outside the class.

Reply:
Agree that we can probably use atomic and cas to avoid synchronization. We're not an asynchronous framework or anything, this is just our export API so we wouldn't expect significant overhead due to extra objects from the atomics

Reply:
I realized my comment about list of runnables may make it impractical to avoid a lock.

Reply:
I can replace the use of synchronised with an internal lock. It isn’t possible to use atomics for each field as we require them to be all synchronised simultaneously.

Reply:
Also, please avoid wording like “queasy” when providing feedback. The use of volatile where I have it is perfectly reasonable. Please be clearer as to why an approach is unsuitable.

Reply:
Sorry.. I was just trying to express that what is implemented here has a mix of different concurrency paradigms in it. When I see that, it is a "code smell". And "code smells" make me uncomfortable. When I read this code, I have a hard time understanding why it is written the way it is, and that's a problem.

Reply:
In essence, using volatile on a boolean is more efficient than using an `AtomicBoolean`. A reason to use `AtomicBoolean` would be to use its CAS operations, which are not required.

Reply:
Could we just replace both fields with a single AtomicReference holding the return code itself? Then you could CAS to set it?

Reply:
> Could we just replace both fields with a single AtomicReference holding the return code itself? Then you could CAS to set it?

They'd be three fields... We could use a CAS on a state object, but I've now pushed some code that uses a reentrant lock. See what you think of that... it is quite focused on reducing object allocations.

Reply:
So, as far as I understand this class (and maybe I'm misunderstanding), there are 3 states. 1) incomplete, 2) complete with failure, 3) complete with success. This can be modeled with one AtomicReference that holds the success/failure. AtomicReference is unset: incomplete, AtomicReference is set to failure: complete with failure, AtomicReference is set to success: complete with success.

Am I missing something?

Reply:
We need to lock both the completion state and the callback list (`actions`) at the same time. Not doing so could lead to a race condition where `action` is added but not executed immediately because the completion has subsequently occurred. You would, therefore, end up with an action never being called.

Reply:
Sure, but even if you need the explicit locking to manage the actions, you can still store the complete state in one nullable field, whether you store it in an AtomicReference or not. Currently, you're using two fields, updates to which need to be coordinated, which is inherently more complex that just managing one field with the result in it.

Reply:
Ah, good point. I have now pushed the change.